snappy/security.go

```diff
@@ -20,94 +20,302 @@
+		newCaps, err := readSingleCapFile(fn)
+		// its ok if the file does not exist
+		if err != nil && os.IsNotExist(err) {
+			continue
+		}
+		// but any other error is not ok
+		if err != nil {
+			return nil, err
+		}
+		p = append(p, newCaps...)
+		found = true
+		break
+	}
 
-	if hasSecPol || hasSecOvr {
-		return nil
+	if found == false {
+		return nil, &errPolicyNotFound{"cap", sp, capName}
+	}
+
+	return p, nil
+}
+
+// findCaps returns the security template content for the given list
+// of security-caps.
+func (sp *securityPolicyType) findCaps(caps []string, templateName string) ([]string, error) {
+	// XXX: this is snappy specific, on other systems like the phone we may
+	// want different defaults.
+	if templateName == "" && caps == nil {
+		caps = defaultPolicyGroups
+	}
+
+	// Nothing to find if caps is empty
+	if len(caps) == 0 {
+		return nil, nil
+	}
+
+	subdir := filepath.Join("policygroups", defaultPolicyVendor(), defaultPolicyVersion())
+	parentDir := filepath.Join(sp.policyDir(), subdir)
+	fwParentDir := filepath.Join(sp.frameworkPolicyDir(), "policygroups")
+
+	var p []string
+	for _, c := range caps {
+		newCap, err := sp.findSingleCap(c, parentDir, fwParentDir)
+		if err != nil {
+			return nil, err
+		}
+		p = append(p, newCap...)
 	}
 
-	// generate apparmor template
-	apparmorJSONFile := m.Integration[hookName]["apparmor"]
-	securityJSONContent, err := s.generateApparmorJSONContent()
+	return p, nil
+}
+
+func defaultPolicyVendor() string {
+	// FIXME: slightly ugly that we have to give a prefix here
+	return fmt.Sprintf("ubuntu-%s", release.Get().Flavor)
+}
+
+// findUbuntuVersion determines the version (eg, 15.04, 15.10, etc) of the
+// system, which is needed for determining the security policy
+// policy-version
+func findUbuntuVersion() (string, error) {
+	content, err := ioutil.ReadFile(lsbRelease)
 	if err != nil {
-		return err
+		logger.Noticef("Failed to read %q: %v", lsbRelease, err)
+		return "", err
 	}
-	if err := ioutil.WriteFile(filepath.Join(buildDir, apparmorJSONFile), securityJSONContent, 0644); err != nil {
-		return err
+
+	for _, line := range strings.Split(string(content), "\n") {
+		if strings.HasPrefix(line, "DISTRIB_RELEASE=") {
+			tmp := strings.Split(line, "=")
+			if len(tmp) != 2 {
+				return "", errSystemVersionNotFound
+			}
+			return tmp[1], nil
+		}
 	}
 
-	return nil
+	return "", errSystemVersionNotFound
+}
+func defaultPolicyVersion() string {
+	// note that we can not use release.Get().Series here
+	// because that will return "rolling" for the development
+	// version but apparmor stores its templates under the
+	// version number (e.g. 16.04) instead
+	ver, err := findUbuntuVersion()
+	if err != nil {
+		// when this happens we are in trouble
+		panic(err)
+	}
+	return ver
+}
+
+const allowed = `abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789`
+
+// Generate a string suitable for use in a DBus object
+func dbusPath(s string) string {
+	dbusStr := ""
```

Comment:
We have some legroom as this isn't super performance sensitive, but ideally for building strings we use either []byte for something trivial, or a bytes.Buffer most of the time. Building it with strings implies paying the cost of strings being immutable.